python_modules/dagster/dagster/_core/definitions/partition_mapping.py

```diff
@@ -248,42 +249,112 @@ def get_downstream_partitions_for_partition_range(
+    a dimension of the MultiPartitionsDefinition.
 
-    For an upstream partition key X, this partition mapping assumes that any downstream
-    multi-partition key with X in the selected dimension is a downstream dependency.
+    This class handles the case where the upstream asset is multipartitioned and the
+    downstream asset is single dimensional, and vice versa.
+
+    For a partition key X, this partition mapping assumes that any multi-partition key with
+    X in the selected dimension is a dependency.
 
     Args:
-        partition_dimension_name (str): The name of the partition dimension in the downstream
-            MultiPartitionsDefinition that matches the upstream partitions definition.
+        partition_dimension_name (Optional[str]): The name of the partition dimension in the
+            MultiPartitionsDefinition that matches the single-dimension partitions definition.
     """
 
-    def __init__(self, partition_dimension_name: str):
-        self.partition_dimension_name = partition_dimension_name
+    def __new__(cls, partition_dimension_name: Optional[str] = None):
+        return super(MultiToSingleDimensionPartitionMapping, cls).__new__(
+            cls,
+            partition_dimension_name=check.opt_str_param(
+                partition_dimension_name, "partition_dimension_name"
+            ),
+        )
 
-    def _check_upstream_partitions_def_equals_selected_dimension(
+    def _check_partitions_defs_and_get_partition_dimension_name(
         self,
         upstream_partitions_def: PartitionsDefinition,
-        downstream_partitions_def: MultiPartitionsDefinition,
-    ) -> None:
-        matching_dimensions = [
-            partitions_def
-            for partitions_def in downstream_partitions_def.partitions_defs
-            if partitions_def.name == self.partition_dimension_name
-        ]
-        if len(matching_dimensions) != 1:
-            check.failed(f"Partition dimension '{self.partition_dimension_name}' not found")
-        matching_dimension_def = next(iter(matching_dimensions))
-
-        if upstream_partitions_def != matching_dimension_def.partitions_def:
+        downstream_partitions_def: PartitionsDefinition,
+    ) -> Tuple[PartitionsDefinition, PartitionsDefinition, str]:
+        if not _can_infer_single_to_multi_partition_mapping(
+            upstream_partitions_def, downstream_partitions_def
+        ):
             check.failed(
-                "The upstream partitions definition does not have the same partitions definition "
-                f"as dimension {matching_dimension_def.name}"
+                "This partition mapping defines a relationship between a multipartitioned and"
+                " single dimensional asset. The single dimensional partitions definition must be a"
+                " dimension of the MultiPartitionsDefinition."
+            )
+
+        multipartitions_def = next(
```

Comment:
Feel free to ignore, but this would arguably a little more standard as an if/else.